pkg/unidling/controller.go

```diff
@@ -0,0 +1,178 @@
+					glog.Errorf("Unable to unidle %s/%s: %v", targetNamespace, targetServiceName, err)
+				}
+			},
+		},
+	)
+
+	unidlingController.controller = controller
+
+	return unidlingController
+}
+
+func (c *UnidlingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go c.controller.Run(stopCh)
+}
+
+// TODO: does this need to send error events?
+func (c *UnidlingController) handleEvent(event *kapi.Event) error {
+	targetNamespace := event.InvolvedObject.Namespace
+	targetServiceName := event.InvolvedObject.Name
+
+	targetEndpoints, err := c.endpointsNamespacer.Endpoints(targetNamespace).Get(targetServiceName)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve service: %v", err)
+	}
+
+	// check when we were idled to prevent historical unidling
+	idledTimeRaw, wasIdled := targetEndpoints.Annotations[unidling.IdledAtAnnotation]
+	if !wasIdled {
+		glog.V(5).Infof("UnidlingController received a NeedPods event for a service that was not idled, ignoring")
+		return nil
+	}
+
+	idledTime, err := time.Parse(time.RFC3339, idledTimeRaw)
+	if err != nil {
+		return fmt.Errorf("unable to check idled-at time: %v", err)
+	}
+	if event.LastTimestamp.Time.Before(idledTime) {
+		glog.V(5).Infof("UnidlingController received an out-of-date NeedPods event, ignoring")
+		return nil
+	}
+
+	// TODO: update this to a CrossVersionObjectReference once we pull in Kube version where apis/autoscaling has it unversioned
+	// TODO: ew, this is unversioned.  Such is life when working with annotations.
+	var targetScalables []kextapi.SubresourceReference
+	if targetScalablesStr, hasTargetScalables := targetEndpoints.Annotations[unidling.UnidleTargetAnnotation]; hasTargetScalables {
+		if err := json.Unmarshal([]byte(targetScalablesStr), &targetScalables); err != nil {
+			glog.Errorf("Unable to unmarshal target scalable references while unidling service %s/%s: %v", targetNamespace, targetServiceName, err)
+			return nil
+		}
+	} else {
+		glog.V(4).Infof("Service %s/%s had no scalables to unidle", targetNamespace, targetServiceName)
+		targetScalables = []kextapi.SubresourceReference{}
+	}
+
+	targetScalablesSet := make(map[kextapi.SubresourceReference]bool, len(targetScalables))
+	for _, v := range targetScalables {
+		targetScalablesSet[v] = true
+	}
+
+	scales := c.scaleNamespacer.Scales(targetNamespace)
+	for _, scalableRef := range targetScalables {
+		scale, err := scales.Get(scalableRef.Kind, scalableRef.Name)
+		if err != nil {
+			glog.Errorf("Unable get scale for %s %q while unidling service %s/%s: %v", scalableRef.Kind, scalableRef.Name, targetNamespace, targetServiceName, err)
+			if errors.IsNotFound(err) {
+				delete(targetScalablesSet, scalableRef)
+			}
+			continue
+		}
+
+		if scale.Spec.Replicas > 0 {
+			glog.Errorf("%s %q is not idle, skipping while unidling service %s/%s", scalableRef.Kind, scalableRef.Name, targetNamespace, targetServiceName)
+			continue
+		}
+
+		scale.Spec.Replicas = 1
```

Comment:
I think we discussed this a bit in the unidling doc, but, IMO, it doesn't make sense to restore to "full" because the scalable has presumably been idled since it hasn't received any traffic (or was very low on some other metric).

If I have 5 replicas, but I've received no traffic in the past 24h, chances are I'm probably not going to see a lot of traffic after wakeup (although I could see a scenario where somebody is aggressively idling, and this is not the case).

If we think waking up to "full" is a thing people are going to want, then it should probably be an option to oc idle, with a default to "wake up to 1".

What do other people think?